core/src/main/java/hudson/ClassicPluginStrategy.java

```diff
@@ -301,9 +302,17 @@ protected ClassLoader createClassLoader(List<File> paths, ClassLoader parent, At
             }
         }
 
-        AntClassLoader classLoader = new AntClassLoader(parent, true);
-        classLoader.addPathFiles(paths);
-        return classLoader;
+        if (useAntClassLoader) {
+            AntClassLoader classLoader = new AntClassLoader(parent, true);
+            classLoader.addPathFiles(paths);
+            return classLoader;
+        } else {
+            List<URL> urls = new ArrayList<>();
+            for (File path : paths) {
+                urls.add(path.toURI().toURL());
+            }
+            return new URLClassLoader2(urls.toArray(new URL[0]), parent);
```

Comment:
if you are using something shared, then yes it will give potentially poorer results as obtaining the list can change between getting the size and then trying to construct.

Reply:
Please read [Arrays of Wisdom of the Ancients](https://shipilev.net/blog/2016/arrays-wisdom-ancients/).